neps/nep-0393.md

```diff
@@ -0,0 +1,570 @@
+
+We recommend that each issuer will use only one registry to avoid complex reconciliation and assure single source of truth.
+
+Moreover, a registry provides an efficient way to query multiple tokens for a single user. This will allow implementation of use cases such us:
+
+- SBT based identities (main use case of the `i-am-human` protocol);
+- SBT classes;
+- decentralized societies.
+
+The registry fills a central and important role. But it is **not centralized**, as anyone can create their own registry and SBT issuers can choose which registry to use. It's also not too powerful, as almost all of the power (mint, revoke, burn, recover, etc) still remains with the SBT issuer and not with the registry.
+
+#### Recovery within an SBT Registry
+
+SBT registry can define it's own mechanism to atomically recover all tokens related to one account and execute soul transfer to another account, without going one by one through each SBT issuer (sometimes that might be even not possible). Below we list few ideas a Registry can use to implement recovery:
+
+- KYC based recovery
+- Social recovery
+
+![Social Recovery, Image via “Decentralized Society”](https://bankless.ghost.io/content/images/public/images/cdb1fc23-6179-44f0-9bfe-e5e5831492f7_1399x680.png)
+
+SBT Registry based recovery is not part of this specification.
+
+#### Minting authorization
+
+A registry can limit which contracts can mint SBTs by implementing a custom issuer registration methods. Example: a simple access control list managed by a DAO.
+
+#### Burning tokens
+
+Registry MAY expose a mechanism to allow an account to burn an unwanted token. The exact mechanism is not part of the standard and it will depend on the registry implementation. We only define a standard `Burn` event, which must be emitted each time a token is removed from existence. Some registries may forbid accounts to burn their tokens in order to preserve specific claims. Ultimately, NEAR is a public blockchain, and even if a token is burned, it's trace will be preserved.
+
+#### Personal Identifiable Information
+
+Issuers must not include any PII into any SBT.
+
+### Smart contract interface
+
+For the Token ID type we propose `u64` rather than `U128`. `u64` capacity is more than 1e19. If we will mint 10'000 SBTs per second, then it will take us 58'494'241 years to fill the capacity.
+Today, the JS integer limit is `2^53-1 ~ 9e15`. Similarly, when minting 10'000 SBTs per second, it will take us 28'561 years to reach the limit. So, we don't need u128 nor the String type. However, if for some reason, we will need to get u64 support for JS, then we can always add another set of methods which will return String, so making it compatible with NFT standard (which is using `U128`, which is a string). Also, it's worth to note, that in 28'000 years JS (if it will still exists) will be completely different.
```

Comment:
You don't make the life of JS folks any simpler. Parsing numbers bigger than $2^{53}$ from JSON payload is painful, so if you say that it is legal to have such big numbers, you force JS folks to use custom JSON parsers as browser parsers will lose the precision on deserialization OR give up on correctness/compliance from the get-go. Why do you want to put developers into this hassle?

I would recommend sticking to a stringified u64, or don't even define what string should contain, or standardizing the usage of u53 / u52.

Reply:
> I want to stick to number, rather than string.

@robert-zaremba Then I would standardize TokenId to be u53, so anything beyond that can be treated as not compliant with the spec rather than a bug in the implementation that would be painful to solve.

> $2^53$ is large enough to cover the usage for next hundreds of years.

This is only true if you assume that the TokenIds will be allocated incrementally. Someone may decide to allocate ranges as IPv4 did, and then suddenly Tokens with 2^60 will be legal and cause problems on JS side.